keras_rs/src/metrics/ranking_metric.py

```diff
@@ -0,0 +1,200 @@
+        """Abstract method, should be implemented by subclasses."""
+        pass
+
+    def update_state(
+        self,
+        y_true: types.Tensor,
+        y_pred: types.Tensor,
+        sample_weight: Optional[types.Tensor] = None,
+    ) -> None:
+        """
+        Accumulates statistics for the ranking metric.
+
+        Args:
+            y_true: tensor. Ground truth values. Of shape `(list_size)`
+                for unbatched inputs or `(batch_size, list_size)` for batched
+                inputs.
+            y_pred: tensor. The predicted values, of shape `(list_size)` for
+                unbatched inputs or `(batch_size, list_size)` for batched
+                inputs. Should be of the same shape as `y_true`.
+            sample_weight: float/tensor. Can be float value, or tensor of
+                shape `(list_size)` or `(batch_size, list_size)`. Defaults to
+                `None`.
+        """
+        # TODO (abheesht): Should `y_true` be a dict, with `"mask"` as one key
+        # for parity  with pairwise losses?
+
+        # === Convert to tensors, if list ===
+        # TODO (abheesht): Figure out if we need to cast tensors to
+        # `self.dtype`.
+        if isinstance(y_true, list):
+            y_true = ops.convert_to_tensor(y_true)
+        if isinstance(y_pred, list):
+            y_pred = ops.convert_to_tensor(y_pred)
+        # `sample_weight` can be a scalar too.
+        if isinstance(sample_weight, (list, float, int)):
+            sample_weight = ops.convert_to_tensor(sample_weight)
+
+        # === Process `sample_weight` ===
+        if sample_weight is None:
+            sample_weight = ops.cast(1, dtype=y_pred.dtype)
+
+        y_true_shape = ops.shape(y_true)
+        y_true_rank = len(y_true_shape)
+        sample_weight_shape = ops.shape(sample_weight)
+        sample_weight_rank = len(sample_weight_shape)
+
+        # Check `y_true_rank` first. Can be 1 for unbatched inputs, 2 for
+        # batched.
+        check_rank(y_true_rank, allowed_ranks=(1, 2), tensor_name="y_true")
+
+        # Check `sample_weight` rank. Should be between 0 and `y_true_rank`.
+        check_rank(
+            sample_weight_rank,
+            allowed_ranks=tuple(range(y_true_rank + 1)),
+            tensor_name="sample_weight",
+        )
+
+        # If `sample_weight` rank is 1, it should be of shape `(batch_size,)`.
+        # Otherwise, it should be of shape `(batch_size, list_size)`.
+        if sample_weight_rank == 1:
+            check_shapes_compatible(sample_weight_shape, (y_true_shape[0],))
+            # Uprank this, so that we get per-list weights here.
+            sample_weight = ops.expand_dims(sample_weight, axis=1)
+        elif sample_weight_rank == 2:
+            check_shapes_compatible(sample_weight_shape, y_true_shape)
+
+        # Want to make sure `sample_weight` is of the same shape as
+        # `y_true`.
+        sample_weight = ops.multiply(ops.ones_like(y_true), sample_weight)
```

Comment:
What is this supposed to do? It is to handle the case when `y_true` is ragged / sparse and sample_weight is dense?